pkg/controller/direct/compute/forwardingrule_controller.go

```diff
@@ -0,0 +1,460 @@
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting ComputeForwardingRule", "name", a.resourceID)
+
+	var err error
+	forwardingRule := &computepb.ForwardingRule{}
+	if a.Location == "global" {
+		req := &computepb.GetGlobalForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Project:        a.projectID,
+		}
+		forwardingRule, err = a.globalForwardingRulesClient.Get(ctx, req)
+	} else {
+		req := &computepb.GetForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Region:         a.Location,
+			Project:        a.projectID,
+		}
+		forwardingRule, err = a.forwardingRulesClient.Get(ctx, req)
+	}
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting ComputeForwardingRule %q failed: %w", a.resourceID, err)
+	}
+	a.actual = forwardingRule
+	return true, nil
+}
+
+func (a *forwardingRuleAdapter) Create(ctx context.Context, u *unstructured.Unstructured) error {
+	if a.projectID == "" {
+		return fmt.Errorf("project is empty")
+	}
+	if a.resourceID == "" {
+		return fmt.Errorf("resourceID is empty")
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating ComputeForwardingRule", "name", a.resourceID)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+
+	forwardingRule := ComputeForwardingRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	forwardingRule.Name = direct.LazyPtr(a.resourceID)
+	forwardingRule.Labels = desired.Labels
+	// TODO(yuhou): handle default spec values
+	if forwardingRule.LoadBalancingScheme == nil {
+		forwardingRule.LoadBalancingScheme = direct.LazyPtr("EXTERNAL")
+	}
+
+	var err error
+	op := &gcp.Operation{}
+	created := &computepb.ForwardingRule{}
+	if a.Location == "global" {
+		req := &computepb.InsertGlobalForwardingRuleRequest{
+			ForwardingRuleResource: forwardingRule,
+			Project:                a.projectID,
+		}
+		getReq := &computepb.GetGlobalForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Project:        a.projectID,
+		}
+		op, err = a.globalForwardingRulesClient.Insert(ctx, req)
+		if err != nil {
+			return fmt.Errorf("creating ComputeForwardingRule %s failed: %w", a.resourceID, err)
+		}
+		log.V(2).Info("successfully created ComputeForwardingRule", "name", a.resourceID)
+		// Get the created resource
+		created, err = a.globalForwardingRulesClient.Get(ctx, getReq)
+		if err != nil {
+			if direct.IsNotFound(err) {
+				return err
+			}
+			return fmt.Errorf("getting ComputeForwardingRule %q failed: %w", a.resourceID, err)
+		}
+	} else {
+		req := &computepb.InsertForwardingRuleRequest{
+			ForwardingRuleResource: forwardingRule,
+			Region:                 a.Location,
+			Project:                a.projectID,
+		}
+		getReq := &computepb.GetForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Region:         a.Location,
```

Comment:
Nit: Why is location exported?

Reply:
`GetForwardingRuleRequest` is regional, Region is required. The usage of `region` and `location` can be confusing, some APIs use `region`, some use `location`. We need a consistent pattern for this. For now, KCC supports `location` only, if API asks for `region` then we use `location` as the value.